### PR TITLE
Serialization for dask

### PR DESCRIPTION
**Pull request by SimonHeybrock**
Proof of concept serialization for dask. This is not intended for merging at this point.

Dependencies:
```
conda install -c conda-forge dask distributed
```

Usage:
```python
from dask.distributed import Client
import scipp as sc


def func(val):
    var = sc.ones(dims=['x', 'y'], shape=[1000, 1000], with_variances=True) * val
    return sc.DataArray(var, coords={'x':sc.linspace(dim='x', start=0.0, stop=1.0, num=1001)})

client = Client()
da = client.map(func, range(10))
from functools import partial, reduce
total = client.submit(partial(reduce, lambda x,y: x+y), da)
total.result()
```

We can also gather the arrays, instead of reducing:
```python
client.gather(da)
```

~Note that serialization only supports types that numpy understands at this point.~

### CODE REVIEW

File: src/scipp/__init__.py [hunk 1]
```diff
@@ -107,3 +107,32 @@
 del _binding
 
 from . import data
+
+try:
+    from typing import List, Dict, Tuple, Union
+    from distributed.protocol import dask_serialize, dask_deserialize
+
+    @dask_serialize.register(Variable)
```
[the remark this answers is not in the document]
Comment (SimonHeybrock) [reply src/scipp/__init__.py:115]: No, actually it should just work now that I am using HDF5IO.

**Comment by SimonHeybrock** [comment]
Here is another example:

```python
from dask.distributed import Client
client = Client()
import scipp as sc
```
```python
chunk_len = 5 * 60


def load_chunk(chunk):
    import scippneutron as scn

    da = scn.load(
        filename=scn.data.get_path("PG3_4844_event.nxs"),
        mantid_args={
            "FilterByTimeStart": chunk * chunk_len,
            "FilterByTimeStop": (chunk + 1) * chunk_len,
        },
    )
    return da

# may need to limit number of workers per node, this takes unusually long
# when loading more chunks on my local machine
raw = client.map(load_chunk, range(4))
```
```python
def reduce(da):
    import scippneutron as scn
    da = scn.convert(da, 'tof', 'dspacing', scatter=True)
    dspacing = sc.linspace(dim='dspacing', start=0.0, stop=2.0, num=1000, unit='Angstrom')
    return sc.histogram(da.bins.concat('spectrum'), bins=dspacing)

da = client.map(reduce, raw)
total = sc.concat(client.gather(da), 'chunk')
total
```

**Comment by jl-wynen** [comment]
Have you checked if scipp respects the number of threads per worker that you set?

**Comment by SimonHeybrock** [comment]
@jl-wynen I ran a small experiment, with a fake workload:

n_workers | time [s]
---|---
1|43
2|30
4|24
8|24
12|25
24|26

My system has 12 cores / 24 threads. I think the result is impressively good, I think it implies that either
a) oversubscribing is beneficial, or
b) TBB finds a good thread count automatically

We would need to run more experiments to draw a conclusion. Possibly more could be gained by tuning the TBB thread-count (depending on the function)?

Code:

```python
import scipp as sc
from dask.distributed import Client
client = Client(threads_per_worker=1, n_workers=24)
client
```
```python
def f(x):
    import scipp as sc
    var = sc.ones(dims=['x'], shape=[10000000])
    for i in range(20):
        sc.asin(var)
    return x
```
```python
%%time
x = client.map(f, range(100))
sum(client.gather(x))
```

**Comment by jl-wynen** [comment]
Before investigating threading, we should measure how much time is spend on computation vs communication. I'm not sure if I would call your results impressively good. Barring cache effects (can they be this big?) and ignoring communication, I would have expected the 2 worker case to be almost twice as fast the 1 worker case.

**Comment by SimonHeybrock** [comment]
> Before investigating threading, we should measure how much time is spend on computation vs communication.

There is no large data communicated in the example.

> I'm not sure if I would call your results impressively good. Barring cache effects (can they be this big?) and ignoring communication, I would have expected the 2 worker case to be almost twice as fast the 1 worker case.

Why? Scipp is multi-threaded so it should be fast even with a single worker (unless the payload size is too small for efficient thread fork/join, I'll run another check for that).

**Comment by SimonHeybrock** [comment]
Ran the same check again, but with 4x larger working set (4x fewer iterations in `f()`): Speedup from 1 to 2 workers is about the same, i.e., a too-small working set it *not* the reason for limited performance.

Maybe it is memory allocation, which can be faster if done in multiple processes?

**Comment by SimonHeybrock** [comment]
For the most part, the difference indeed appears to be faster multi-process memory allocation. If we test a workload with much fewer allocations:

```python
def f(x):
    import scipp as sc
    var = sc.ones(dims=['x'], shape=[10000000])
    for i in range(80):
        var += var
    return x
```

we get:

n_workers | time [s]
---|---
1|53
2|46
4|43

Still an improvement with multiple processes, but much less.

**Comment by SimonHeybrock** [comment]
I have cleaned this up, so it is ready for an actual review. Testing is added, but not with dask itself.

**Review by nvaytet — APPROVED** [review]
looks fantastically simple ;-)

**Pull request merged**